### ISSUE TITLE
Loading modules multiple times

### ISSUE THREAD

**Issue by qaisjp** [multitheftauto/mtasa-blue#1486]
**Describe the bug**

You can load the same module multiple times just by omitting the `.dll` extension.

**To reproduce**

1. `loadmodule modname.dll` (loaded!)
2. `loadmodule modname` (loaded!)
3. `run #getLoadedModules() == 2`

**Expected behaviour**

1. `loadmodule modname.dll` (loaded!)
2. `loadmodule modname` (already loaded!)
3. `run #getLoadedModules() == 1`

**Screenshots**

![image](https://user-images.githubusercontent.com/923242/83383951-19d9e900-a3de-11ea-9b35-806ceee6d2cc.png)

**Version**

1.5.7 latest master 8c373a478a060daa26cc5b9a477b74bfc1a7aaa3
**Additional context**

**Comment by qaisjp** [comment]
After investigating this and lazily working on a fix, it doesn't look like a very straightforward fix (or maybe there is one - I don't know the Windows API well.)

I had the idea to canonicalise to the real filepath (and basename) using [GetModuleFileNameA](https://docs.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-getmodulefilenamea) and ([`std::filesystem::path::filename`](https://en.cppreference.com/w/cpp/filesystem/path/filename), to get the basename), but this doesn't work in the below scenario or any scenario where the user's original input is not the canonical filepath.

![image](https://user-images.githubusercontent.com/923242/83385128-8f46b900-a3e0-11ea-84bb-6a77df3eeedc.png)

<sup><strong>Caption:</strong> So the user's input is taken as the unique short filename, which is used for comparisons.</sup>

Now the issue is that we can only get that canonical filepath _after_ having loaded the module, so it's not good enough. (Rejecting & unloading the module this late would be possible, but it's still pretty messy.)

Is anyone aware of a way to resolve the filename before calling `LoadLibrary`?

**Comment by qaisjp** [comment]
So there's basically a StackOverflow post about this ["How can I find the full path of a DLL without actually calling LoadLibrary?"](https://stackoverflow.com/questions/17713251/how-can-i-find-the-full-path-of-a-dll-without-actually-calling-loadlibrary).

The key issue here, highlighted by that post, is described in the _Remarks_ section of the [SearchPathA](https://docs.microsoft.com/en-gb/windows/win32/api/processenv/nf-processenv-searchpatha?redirectedfrom=MSDN) documentation:

> The `SearchPath `function is not recommended as a method of locating a .dll file if the intended use of the output is in a call to the [`LoadLibrary`](https://docs.microsoft.com/windows/desktop/api/libloaderapi/nf-libloaderapi-loadlibrarya) function. This can result in locating the wrong .dll file because the search order of the `SearchPath` function differs from the search order used by the `LoadLibrary` function. If you need to locate and load a .dll file, use the `LoadLibrary` function.

Questions are:

1. Should we just ignore the recommendation, and SearchPath -> check path clash -> LoadLibrary path directly.
2. Is this issue even important enough to be breaking those recommendations?

**Comment by AlexTMjugador** [comment]
The [`LoadLibraryExA` documentation](https://docs.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-loadlibraryexa) says this (of course, `LoadLibraryExW` too):

> If the string specifies a module name without a path and the file name extension is omitted, the function appends the default library extension .dll to the module name. To prevent the function from appending .dll to the module name, include a trailing point character (.) in the module name string.

Therefore, if we want to avoid the behaviour of appending the extension automatically, that can be used.

On a side note:

> If the string specifies a module name without a path and more than one loaded module has the same base name and extension, the function returns a handle to the module that was loaded first.

So, as I see it, if we guarantee that the parameter passed to `LoadLibraryEx` is not a path (just a module name), we don't have to worry about loading the module twice.

Edit: if for some reason the above ideas are not enough for a proper fix, I guess we could use `GetModuleHandleA/W` to check if a module is loaded after canonicalizing its path and before actually loading it, but that's racey unless we put that in a critical section.

**Comment by qaisjp** [comment]
Hmm I thought our use of LoadLibrary looks in multiple locations, but according to the Windows docs:

> If the string specifies a full path, the function searches only that path for the module.

Which testing with an official build vs. my self-built server, for some reason the `.dll` appending ALWAYS happens when self-built (Debug), but never happens with an official (Release) build.

Which means that we can just normalise the filepath (to get rid of `./`) and always add a trailing `.` (because adding the dot won't break BC).

Technical point: we use [LoadLibrary](https://docs.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-loadlibrarya) (A, because I don't think we set UNICODE), but the same information applies for both of them.

**Comment by AlexTMjugador** [comment]
I looked a bit into the source code related to module loading, more precisely [this method](https://github.com/multitheftauto/mtasa-blue/blob/4e1629f2cc3af819b5a66f13d952c126bea83de7/Server/mods/deathmatch/logic/CConsoleCommands.cpp#L1320), and I can see that `szArguments` is passed around without any sanitizing to `CLuaModuleManager::LoadModule`, which interprets that as a short file name. The short file name is, in turn, the distinctive module name, as returned by `_GetName()`, and the module name is used [here](https://github.com/multitheftauto/mtasa-blue/blob/927509565ab3a8d48df006440654c38e0942aa89/Server/mods/deathmatch/logic/lua/CLuaModuleManager.cpp#L64) to check if the same module was already loaded.

Therefore, the way I see it, a fix could be canonicalizing the `szArguments` value, appending the extension if not provided, and then take just the basename of the canonicalized path. So, for instance, something like `.\hello` would be canonicalized to `C:\Whatever\Is\The\Working\Dir\hello.dll`, from which we extract the basename `hello.dll`. And a similar string like `hello.dll` would also canonicalize to the same basename, so the check for whether a module with the same name was already loaded would work no matter what the user input is. On the other hand, because no documentation ever specified this precisely how the arguments to `loadmodule` would be resolved to an actual library, I think that backwards compatibility should not be an issue, and I don't really think that much scripts out there rely on the current de facto behaviour for this command.

Note that the actual path passed to `LoadLibrary` would not be modified by the changes I discussed. It is generated [here](https://github.com/multitheftauto/mtasa-blue/blob/4e1629f2cc3af819b5a66f13d952c126bea83de7/Server/mods/deathmatch/logic/CConsoleCommands.cpp#L1332) in a relatively naïve way, as it just ends up concatenating strings without making sure that they end up in the expected module directory, which `CLuaModule::_LoadModule` then [happily accepts](https://github.com/multitheftauto/mtasa-blue/blob/927509565ab3a8d48df006440654c38e0942aa89/Server/mods/deathmatch/logic/lua/CLuaModule.cpp#L73) and is admittely a bit of an unexpected feature. But that is a matter for another issue, perhaps.

So what do you think about this?

TL;DR: for solving this issue, I think that making sure that `szArguments` is always canonicalized to the same string for the same module file, no matter the arguments for `loadmodule`, is enough.

**Comment by qaisjp** [comment]
> appending the extension if not provided

I think we can just append a dot in all cases, instead of appending an extension, as the auto-append feature seems to only happen in Debug builds, and not Release builds. So appending a dot will just make both builds behave consistently.

> from which we extract the basename `hello.dll`.

We don't necessarily want just the basename, as the user might want to do `subfolder/mod.dll`. 

> as it just ends up concatenating strings without making sure that they end up in the expected module directory

Accessing the parent folder (i.e. `loadmodule ../../../test.dll`) is blocked at point of `IsValidFilepath(szArguments)`.

> TL;DR: for solving this issue, I think that making sure that `szArguments` is always canonicalized to the same string for the same module file, no matter the arguments for `loadmodule`, is enough.

Agreed.  I suggest using `::path::lexically_normal` from `std::filesystem` to canonicalise the filename. i.e. converting `./king.dll` to `king.dll.`

We may as well also abandon the use of `IsValidFilepath` + `PathJoin` in the rest of `CConsoleCommands::LoadModule` as well.

Is this something you would like to submit a PR for?

**Comment by AlexTMjugador** [comment]
> I think we can just append a dot in all cases, instead of appending an extension, as the auto-append feature seems to only happen in Debug builds, and not Release builds. So appending a dot will just make both builds behave consistently.

I was referring to the `szArguments` canocalization. At that stage, I find it unnecessary to rely on `LoadLibrary` or `dlopen` to do anything funky with the extension. But anyway that sounds like a nice idea for fixing the behaviour you describe of extension appending not working consistently.

> We don't necessarily want just the basename, as the user might want to do `subfolder/mod.dll`.

Totally right - I didn't consider the possibility of subfolders. Good point.

> Accessing the parent folder (i.e. `loadmodule ../../../test.dll`) is blocked at point of `IsValidFilepath(szArguments)`.

Hmm, indeed it should. But the IsValidFilePath function used there is defined [here](https://github.com/multitheftauto/mtasa-blue/blob/9653039ca8b3f256c0bb545ea98a61e07eabc77a/Shared/mods/deathmatch/logic/Utils.cpp#L502), right? If that's the case, I fail to see how that function makes sure that the path doesn't end up pointing to an arbitrary file the MTA server process has access to; it just makes sure that the path doesn't contain strange characters.

> Agreed. I suggest using `::path::lexically_normal` from `std::filesystem` to canonicalise the filename. i.e. converting `./king.dll` to `king.dll.`
> 
> We may as well also abandon the use of `IsValidFilepath` + `PathJoin` in the rest of `CConsoleCommands::LoadModule` as well.

Sounds like a plan. Although not necessarily bad by themselves, the use of these functions as they are right now scares me a bit for the reason I stated above.

> Is this something you would like to submit a PR for?

I'd love to, but sadly I don't enough free time right now to set up a development environment and do things properly. That doesn't stop me from trying to be helpful, though, so go ahead with this if you want 😄

**Comment by qaisjp** [comment]
> Hmm, indeed it should. But the IsValidFilePath function used there is defined [here](https://github.com/multitheftauto/mtasa-blue/blob/9653039ca8b3f256c0bb545ea98a61e07eabc77a/Shared/mods/deathmatch/logic/Utils.cpp#L502), right? If that's the case, I fail to see how that function makes sure that the path doesn't end up pointing to an arbitrary file the MTA server process has access to; it just makes sure that the path doesn't contain strange characters.
> 

IsValidFilePath is called against szArguments:

https://github.com/multitheftauto/mtasa-blue/blob/4e1629f2cc3af819b5a66f13d952c126bea83de7/Server/mods/deathmatch/logic/CConsoleCommands.cpp#L1327-L1332

And IsValidFilePath ensures current character and next character are not both dots:

https://github.com/multitheftauto/mtasa-blue/blob/9653039ca8b3f256c0bb545ea98a61e07eabc77a/Shared/mods/deathmatch/logic/Utils.cpp#L519-L520

Therefore szArguments can "only enter subfolders, and not parent folders", if that makes sense.

The append happens _after_ szArguments is checked.

> I'd love to, but sadly I don't enough free time right now to set up a development environment and do things properly. That doesn't stop me from trying to be helpful, though, so go ahead with this if you want 😄

Okay, I'll do it then :)
